src/CalcViewModel/Common/TraceLogger.cpp

```diff
@@ -236,782 +121,215 @@ namespace CalculatorApp
+            it->count++;
         }
         else
         {
-            radixButtonUsage[windowId]++;
-        }
-        if ((radixButtonUsage[windowId] == 2) && !radixButtonLoggedInSession)
-        {
-            LoggingFields fields{};
-            fields.AddUInt32(L"WindowId", windowId);
-            LogTelemetryEvent(EVENT_NAME_RADIX_BUTTON_USED, fields);
-
-            radixButtonLoggedInSession = true;
+            NumbersAndOperatorsEnum button = safe_cast<NumbersAndOperatorsEnum>(buttonId);
+            buttonLog.push_back(ButtonLog(buttonId, button.ToString()->Data(), mode));
         }
     }
 
-    void TraceLogger::LogAngleButtonUsed(int windowId)
+    void TraceLogger::UpdateWindowCount(size_t windowCount)
     {
-        if (angleButtonUsage.find(windowId) == angleButtonUsage.end())
-        {
-            angleButtonUsage.insert(pair<int, int>(windowId, 1));
-        }
-        else
-        {
-            angleButtonUsage[windowId]++;
-        }
-        if ((angleButtonUsage[windowId] == 2) && !angleButtonLoggedInSession)
+        if (windowCount == 0)
         {
-            LoggingFields fields{};
-            fields.AddUInt32(L"WindowId", windowId);
-            LogTelemetryEvent(EVENT_NAME_ANGLE_BUTTONS_USED, fields);
-
-            angleButtonLoggedInSession = true;
+            currentWindowCount--;
+            return;
         }
+        currentWindowCount = windowCount;
     }
 
-    void TraceLogger::LogFunctionUsage(int windowId)
+    void TraceLogger::LogButtonUsage()
     {
         if (!GetTraceLoggingProviderEnabled())
             return;
 
-        for (int i = 0; i < functionCount; i++)
-        {
-            // log only those functions which are used
-            if (funcLog[i].count > 0)
-            {
-                LoggingFields fields{};
-                fields.AddString(L"FunctionName", funcLog[i].funcName.data());
-                fields.AddUInt32(L"UsageCount", funcLog[i].count);
-                fields.AddUInt32(L"WindowId", windowId);
-                LogTelemetryEvent(EVENT_NAME_FUNCTION_USAGE, fields);
-            }
-        }
-    }
-
-    void TraceLogger::LogHypButtonUsed(int windowId)
-    {
-        if (!isHypButtonLogged)
+        for (auto i : buttonLog)
         {
             LoggingFields fields{};
-            fields.AddUInt32(L"WindowId", windowId);
-            LogTelemetryEvent(EVENT_NAME_HYP_BUTTON_USED, fields);
-
-            isHypButtonLogged = true;
+            fields.AddInt32(L"CalcMode", NavCategory::Serialize(i.mode));
+            fields.AddUInt32(L"ButtonId", i.buttonId);
+            fields.AddString(L"ButtonName", i.buttonName.data());
+            fields.AddUInt32(L"ViewModeId", NavCategory::Serialize(i.mode));
+            fields.AddUInt32(L"UsageCount", i.count);
+            fields.AddUInt64(PDT_PRIVACY_DATA_TAG, PDT_PRODUCT_AND_SERVICE_USAGE);
+            LogLevel2Event(EVENT_NAME_BUTTON_USAGE, fields);
```

Comment:
>LogLevel2Event(EVENT_NAME_BUTTON_USAGE, fields); [](start = 12, length = 48)

It would be better to log a single event with many fields (B5_Count, B5_Mode, etc.) rather than writing dozens of events in a loop. That way we get atomic behavior--either we log usage for all the buttons or none at all.

Reply:
Right now, we are only logging an event for each unique button and mode combination used in the session. Are you suggesting that we dynamically add fields to log all of the buttons in 1 event?

Reply:
Yes, exactly that.

Reply:
I updated the LogButtonUsage method to write the ButtonLog array into a string that uses | and , to separate the fields and button info. So it will write a string like "Mode|Button1|Count,Mode|Button2|count, ..."